android/jni/app-android.cpp

```diff
@@ -29,7 +36,351 @@
+			flags = TOUCH_WHEEL;
+			break;
+		}
+
+		if (flags != 0) {
+			int tool = AMotionEvent_getToolType(input_event, i);
+			flags |= tool << 10;
+
+			TouchInput touch;
+			touch.id = pid;
+			touch.x = AMotionEvent_getX(input_event, i) * dp_xscale;
+			touch.y = AMotionEvent_getY(input_event, i) * dp_yscale;
+			touch.flags = flags;
+			if (touch.flags & TOUCH_DOWN) {
+				input_state.pointer_down[pid] = true;
+			}
+			if (touch.flags & TOUCH_UP) {
+				input_state.pointer_down[pid] = false;
+			}
+
+			retval = NativeTouch(touch) || retval;
+
+			if (pid >= MAX_POINTERS) {
+				ELOG("Too many pointers: %i", pid);
+				return false;	// We ignore 8+ pointers entirely.
+			}
+
+			input_state.pointer_x[pid] = touch.x;
+			input_state.pointer_y[pid] = touch.y;
+			input_state.mouse_valid = true;
+		}
+	}
+
+	return retval;
+}
+
+bool ProcessButton(AInputEvent *input_event, int deviceId) {
+	KeyInput keyInput;
+	keyInput.deviceId = deviceId;
+	keyInput.keyCode = AKeyEvent_getKeyCode(input_event);
+	switch (AKeyEvent_getAction(input_event)) {
+	case AKEY_EVENT_ACTION_DOWN:
+		keyInput.flags = KEY_DOWN;
+		break;
+	case AKEY_EVENT_ACTION_UP:
+		keyInput.flags = KEY_UP;
+		break;
+	default:
+		return false;
+	}
+	if ((keyInput.flags & KEY_DOWN) && AKeyEvent_getRepeatCount(input_event) > 0) {
+		keyInput.flags |= KEY_IS_REPEAT;
+	}
+	return NativeKey(keyInput);
+}
+
+bool ProcessKey(AInputEvent *input_event) {
+	if (AKeyEvent_getKeyCode(input_event) == AKEYCODE_BACK) {
+		if (NativeIsAtTopLevel()) {
+			ILOG("Passing back button press through to Android!");
+			return false;
+		} else {
+			ProcessButton(input_event, DEVICE_ID_DEFAULT);
+			return true;
+		}
+	}
+	return ProcessButton(input_event, DEVICE_ID_KEYBOARD);
+}
+
+bool ProcessJoystickMotion(AInputEvent *joy_event) {
+	AxisInput axis;
+	axis.deviceId = DEVICE_ID_PAD_0;
+
+	return false;
+}
+
+bool ProcessTouchpadMotion(AInputEvent *touchpad_event) {
+	// Xperia Play uses this.
+	ILOG("Xperia Play Touchpad Motion");
+	return false;
+}
+
+// Process the next input event. We can return 1 to prevent Android from further processing the event,
+// such as when catching the back button.
+static int32_t engine_handle_input(struct android_app* app, AInputEvent* input_event) {
+	struct engine* engine = (struct engine*)app->userData;
+	int source = AInputEvent_getSource(input_event) & AINPUT_SOURCE_CLASS_MASK;
+	bool handled = false;
+	if (AInputEvent_getType(input_event) == AINPUT_EVENT_TYPE_MOTION) {
+		if (source & AINPUT_SOURCE_CLASS_JOYSTICK) {
+			handled = ProcessJoystickMotion(input_event);
+		} else if (source & AINPUT_SOURCE_CLASS_POINTER) {
+			handled = ProcessPointerMotion(input_event);
+		} else if (source & AINPUT_SOURCE_TOUCHPAD) {
+			handled = ProcessTouchpadMotion(input_event);
+		}
+	} else if (AInputEvent_getType(input_event) == AINPUT_EVENT_TYPE_KEY) {
+		if (source & (AINPUT_SOURCE_CLASS_JOYSTICK)) {
+			handled = ProcessButton(input_event, DEVICE_ID_PAD_0);
```

Comment:
I wonder if we should (later) map using `AInputEvent_getDeviceId` to get different pad IDs - in case of multiple controllers connected to a single device (e.g. SHIELD, etc.)

-[Unknown]